### PR TITLE
Update to v15.0.0

### PR DESCRIPTION
**Pull request by raulcd**
Checklist
* [x] Used a [personal fork of the feedstock to propose changes](https://conda-forge.org/docs/maintainer/updating_pkgs.html#forking-and-pull-requests)
* [ ] Bumped the build number (if the version is unchanged)
* [x] Reset the build number to `0` (if the version changed)
* [x] [Re-rendered]( https://conda-forge.org/docs/maintainer/updating_pkgs.html#rerendering-feedstocks ) with the latest `conda-smithy` (Use the phrase <code>@<space/>conda-forge-admin, please rerender</code> in a comment in this PR for automated rerendering)
* [x] Ensured the license file is being packaged.

### CODE REVIEW

**Comment by xhochy** [comment]
Please add `14.x` to https://github.com/conda-forge/arrow-cpp-feedstock/blob/482b07558e4c4a70033afb1b68675222918a6f0d/conda-forge.yml#L5

File: recipe/meta.yaml
Comment (h-vetinari) [inline recipe/meta.yaml:27]: Please update the testing submodule, to the commit used as of https://github.com/apache/arrow/tree/apache-arrow-15.0.0

File: conda-forge.yml
Comment (h-vetinari) [inline conda-forge.yml:6]: JFYI, we'll still need to create that branch; I can do this once we merge.

Comment (xhochy) [reply conda-forge.yml:6]: Already did this as I expect we will merge this PR quite soon(ish).

File: recipe/meta.yaml
Comment (raulcd) [reply recipe/meta.yaml:27]: Thanks! I didn't even notice we had that version there. I am learning a lot :)

Comment (h-vetinari) [reply recipe/meta.yaml:27]: The reason is pretty straightforward - we want to run the test suite, which requires the testing submodule, but it is not part of the tarball, so we need to pull it in (and keep it in sync) manually :)

**Comment by raulcd** [comment]
Current failures for pandas tests are being tracked here on Arrow: https://github.com/apache/arrow/issues/39732

**Comment by h-vetinari** [comment]
> Current failures for pandas tests are being tracked here on Arrow: [apache/arrow#39732](https://github.com/apache/arrow/issues/39732)

You could add a pin of `pandas <2.2` to the test requirements (or, if you're worried about how many users will run into this, directly to the run requirements). That way the release of 15.0 becomes independent of backporting the fixes still under development

**Comment by raulcd** [comment]
I am quite confused. I force pushed to force a rebuild due to a timeout and now almost all CI jobs are failing. I'll continue investigating what is going on.

**Comment by h-vetinari** [comment]
Something is going wrong with getting the libs into place:
```
Traceback (most recent call last):
  File "/home/conda/feedstock_root/build_artifacts/apache-arrow_1706091297674/test_tmp/run_test.py", line 11, in <module>
    import pyarrow.flight
  File "/home/conda/feedstock_root/build_artifacts/apache-arrow_1706091297674/_test_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_pla/lib/python3.10/site-packages/pyarrow/flight.py", line 67, in <module>
    raise ImportError(
ImportError: The pyarrow installation is not built with support for 'flight' (libgrpc++.so.1.60: cannot open shared object file: No such file or directory)
WARNING: Tests failed for pyarrow-15.0.0-py310h9a2f4d7_0_cuda.conda - moving package to /home/conda/feedstock_root/build_artifacts/broken
TESTS FAILED: pyarrow-15.0.0-py310h9a2f4d7_0_cuda.conda
```

Also note that you can restart CI from the checks tab (or after clicking on the details of the CI summary) - just please only click once, and only once all other jobs have finished. The interface between GitHub and azure is buggy and it's possible to unintentionally start several runs simultaneously

**Comment by raulcd** [comment]
@conda-forge-admin, please rerender

**Comment by raulcd** [comment]
I think the problem is with the migration for libgrpc and libprotobuf. I am trying to re-render locally with `conda smithy rerender` as CI is failing but I get the same error:
```
INFO:conda_smithy.configure_feedstock:__pycache__ rendering is skipped
INFO:conda_smithy.configure_feedstock:README rendering is skipped
INFO:conda_smithy.configure_feedstock:libthrift0190.yaml from feedstock is ignored and upstream version is used
INFO:conda_smithy.configure_feedstock:libgrpc159_libprotobuf4244.yaml from feedstock is ignored and upstream version is used
INFO:conda_smithy.configure_feedstock:python312.yaml from feedstock is ignored and upstream version is used
INFO:conda_smithy.configure_feedstock:libabseil20230802_libgrpc157_libprotobuf4234.yaml from feedstock is ignored and upstream version is used
WARNING: Setting build platform. This is only useful when pretending to be on another platform, such as for rendering necessary dependencies on a non-native platform. I trust that you know what you're doing.
WARNING: Setting build arch. This is only useful when pretending to be on another arch, such as for rendering necessary dependencies on a non-native arch. I trust that you know what you're doing.
WARNING: No numpy version specified in conda_build_config.yaml.  Falling back to default numpy value of 1.22
Adding in variants from internal_defaults
Adding in variants from /home/raulcd/.cache/conda-smithy/conda_build_config.yaml
INFO:conda_smithy.configure_feedstock:Applying migrations: /home/raulcd/code/arrow-cpp-feedstock/.ci_support/migrations/cuda120.yaml,/home/raulcd/.cache/conda-smithy/share/conda-forge/migrations/libabseil20230802_libgrpc157_libprotobuf4234.yaml,/home/raulcd/.cache/conda-smithy/share/conda-forge/migrations/libthrift0190.yaml,/home/raulcd/.cache/conda-smithy/share/conda-forge/migrations/python312.yaml,/home/raulcd/.cache/conda-smithy/share/conda-forge/migrations/libgrpc159_libprotobuf4244.yaml
Traceback (most recent call last):
  File "/home/raulcd/miniconda3/bin/conda-smithy", line 10, in <module>
    sys.exit(main())
  File "/home/raulcd/miniconda3/lib/python3.9/site-packages/conda_smithy/cli.py", line 724, in main
    args.subcommand_func(args)
  File "/home/raulcd/miniconda3/lib/python3.9/site-packages/conda_smithy/cli.py", line 584, in __call__
    self._call(args, tmpdir)
  File "/home/raulcd/miniconda3/lib/python3.9/site-packages/conda_smithy/cli.py", line 589, in _call
    configure_feedstock.main(
  File "/home/raulcd/miniconda3/lib/python3.9/site-packages/conda_smithy/configure_feedstock.py", line 2552, in main
    render_azure(env, config, forge_dir, return_metadata=True)
  File "/home/raulcd/miniconda3/lib/python3.9/site-packages/conda_smithy/configure_feedstock.py", line 1520, in render_azure
    return _render_ci_provider(
  File "/home/raulcd/miniconda3/lib/python3.9/site-packages/conda_smithy/configure_feedstock.py", line 688, in _render_ci_provider
    migrated_combined_variant_spec = migrate_combined_spec(
  File "/home/raulcd/miniconda3/lib/python3.9/site-packages/conda_smithy/configure_feedstock.py", line 621, in migrate_combined_spec
    combined_spec = variant_add(combined_spec, migration)
  File "/home/raulcd/miniconda3/lib/python3.9/site-packages/conda_smithy/variant_algebra.py", line 290, in variant_add
    return VARIANT_OP[operation](v1, v2)
  File "/home/raulcd/miniconda3/lib/python3.9/site-packages/conda_smithy/variant_algebra.py", line 177, in op_variant_key_add
    new_keys = variant_key_set_union(
  File "/home/raulcd/miniconda3/lib/python3.9/site-packages/conda_smithy/variant_algebra.py", line 115, in variant_key_set_union
    return sorted(out_v, key=partial(_version_order, ordering=ordering))
  File "/home/raulcd/miniconda3/lib/python3.9/site-packages/conda_smithy/variant_algebra.py", line 66, in _version_order
    return ordering.index(v)
ValueError: '11.8' is not in list
```
The only thing I can see when doing the migration with the `11.8` is on the cuda migration:
```
$ grep -R "11.8"  /home/raulcd/.cache/conda-smithy/share/conda-forge/migrations/ 
/home/raulcd/.cache/conda-smithy/share/conda-forge/migrations/cuda120.yaml:      # case: CUDA 11.8
/home/raulcd/.cache/conda-smithy/share/conda-forge/migrations/cuda120.yaml:      - quay.io/condaforge/linux-anvil-cuda:11.8          # [linux64 and os.environ.get("BUILD_PLATFORM") == "linux-64"]
/home/raulcd/.cache/conda-smithy/share/conda-forge/migrations/cuda120.yaml:      - quay.io/condaforge/linux-anvil-ppc64le-cuda:11.8  # [ppc64le and os.environ.get("BUILD_PLATFORM") == "linux-ppc64le"]
/home/raulcd/.cache/conda-smithy/share/conda-forge/migrations/cuda120.yaml:      - quay.io/condaforge/linux-anvil-aarch64-cuda:11.8  # [aarch64 and os.environ.get("BUILD_PLATFORM") == "linux-aarch64"]
/home/raulcd/.cache/conda-smithy/share/conda-forge/migrations/cuda120.yaml:      - quay.io/condaforge/linux-anvil-cuda:11.8          # [ppc64le and os.environ.get("BUILD_PLATFORM") == "linux-64"]
/home/raulcd/.cache/conda-smithy/share/conda-forge/migrations/cuda120.yaml:      - quay.io/condaforge/linux-anvil-cuda:11.8          # [aarch64 and os.environ.get("BUILD_PLATFORM") == "linux-64"]
/home/raulcd/.cache/conda-smithy/share/conda-forge/migrations/cuda120.yaml:      - 11.8                       # [(linux or win64) and os.environ.get("CF_CUDA_ENABLED", "False") == "True"]
```
I am not entirely sure how this migrations work and how `conda-smithy` operates. I can try and manually modify it but it probably is better if someone can take a look into it.

**Comment by raulcd** [comment]
After some debugging on `conda-smithy` I realised I had to manually add:
```
diff --git a/.ci_support/migrations/cuda120.yaml b/.ci_support/migrations/cuda120.yaml
index 17bd59c..6fa3d92 100644
--- a/.ci_support/migrations/cuda120.yaml
+++ b/.ci_support/migrations/cuda120.yaml
@@ -44,6 +44,7 @@ __migrator:
       - 11.0                       # [(linux64 or win) and os.environ.get("CF_CUDA_ENABLED", "False") == "True"]
       - 11.1                       # [(linux64 or win) and os.environ.get("CF_CUDA_ENABLED", "False") == "True"]
       - 11.2                       # [(linux64 or win) and os.environ.get("CF_CUDA_ENABLED", "False") == "True"]
+      - 11.8
       - 12.0                       # [(linux64 or win) and os.environ.get("CF_CUDA_ENABLED", "False") == "True"]
   commit_message: |
     Rebuild for CUDA 12
```
before running the re-render. That's probably a bug on `conda-smithy` when reordering if the version was not there previously.

**Comment by xhochy** [comment]
This is a bit confusing as in this case `conda-smithy` should have pulled the updated migration file from conda-forge-pinning. There, this is already fixed.

File: recipe/meta.yaml
Comment (h-vetinari) [inline recipe/meta.yaml:797]: Is `test_dlpack` related to CUDA? It doesn't [look](https://github.com/apache/arrow/blob/apache-arrow-15.0.0/python/pyarrow/tests/test_dlpack.py#L70) like it at first glance.

If not, please put it on a separate line, with a separate comment why that skip is there. If the skip is not absolutely unavoidable (like not having a GPU in CI forces us to skip CUDA tests), it should go below `# vvvvvvv TESTS THAT SHOULDN'T HAVE TO BE SKIPPED vvvvvvv`.

File: conda-forge.yml
Comment (h-vetinari) [reply conda-forge.yml:6]: > Already did this as I expect we will merge this PR quite soon(ish).

FWIW, let's please only create branches once the PR gets merged. I've now backported #1292 from `main` to `14.x`, but that's kind of annoying to keep track of.

**Comment by h-vetinari** [comment]
> Current failures for pandas tests are being tracked here on Arrow: https://github.com/apache/arrow/issues/39732

I hadn't found this issue when I wrote efd67842e74d241b480d15bf807dea304a1f054a (forgot to check the closed ones), but it would be good to add a reference to that issue above the `pandas <2.2` to remind us that we can drop that for arrow 16.

Similarly for `pytest <8`, https://github.com/apache/arrow/issues/39849 didn't exist yet when I wrote 985aeec6f23abf6d488de33af72acc338346f5cb - could you please add a reference to that in a comment above as well?

File: recipe/meta.yaml
Comment (raulcd) [reply recipe/meta.yaml:797]: The test does require a GPU, I got a failure about that on CI, @alenkaf am I right here?

Comment (AlenkaF) [reply recipe/meta.yaml:797]: The last test from `test_dlpack.py` is related to CUDA: [test_dlpack_cuda_not_supported](https://github.com/apache/arrow/blob/a61f4af724cd06c3a9b4abd20491345997e532c0/python/pyarrow/tests/test_dlpack.py#L124C5-L124C35). But it should be skipped if CUDA is not installed, though.

Comment (raulcd) [reply recipe/meta.yaml:800]: oh! so maybe I can try skipping just that test and run the others. Let me try that.

**Comment by h-vetinari** [comment]
Hey @raulcd, we can either merge this PR or #1297 first. It doesn't really matter all that much TBH. If you want to keep the rebasing effort low, we can merge this to main and #1297 to `14.x`.

Any preferences?

**Comment by raulcd** [comment]
> we can merge this to main and #1297 to `14.x`.
> 
> Any preferences?

I am ok with both. We can merge this to main and #1297 to `14.x` as suggested. I am just waiting for a final CI pass here and I think we can merge.

Thanks a lot for your help while I am learning about conda :)

**Review by h-vetinari — APPROVED** [review]
Thanks for sticking through this!

There were a couple of pretty rare & non-obvious issues, normally it's less of an effort. ;-)

**Pull request merged**
